### ISSUE TITLE
SKDocument.CreateXps() returns null in a .NET Core App

### ISSUE THREAD

**Issue by 8** [mono/SkiaSharp#463]
Hi,

Creating an Xps Document fails when run in a dotnet core app (the returned document instance is null), while the same code works with the full .NET Framework.

```csharp
using (var stream = new SKFileWStream("test.xps"))
using (var document = SKDocument.CreateXps(stream))
{
 // document = null if run from a dotnet core 2.0 app!
}
```

I am using the newest SkiaSharp nuget: 1.60.0

(FYI: `SKDocument.CreatePdf()` works fine from a dotnet core app.)
As the Xps Backend is provided by Skia I would expect it to work in SkiaSharp as well.

Take care,
Martin

**Comment by mattleibow** [comment]
I will check it out, but I just want to confirm that this is running on Windows?

**Comment by 8** [comment]
Hi,
yes, it's a Win10 machine.

It's not a show stopper as the Xps Backend is marked as Experimental anyway and there is the Full .NET Framework workaround.

If you want I can try to investigate myself and check out the source code, maybe I'll find something.
Take care,
Martin

**Comment by 8** [comment]
I've just read the source code and it seems that the Xps Backend is only supported on Windows, as it is based on a Windows COM Object. That's why it's hidden behind a `#ifdef SK_BUILD_FOR_WIN` compiler flag. That's probably why it doesn't run on dotnet core.
Take care,
Martin

**Comment by mattleibow** [comment]
I need to investigate this a bit more. It _should_ work as the .NET Core app is using the same native library as the Full .NET Framework. I am not sure why there is this difference at runtime. The .NET Core unit tests pass, and that creates both a PDF and XPS document.

**Comment by mattleibow** [comment]
I managed to find the issue.
It was the fact that the COM was not initialized. With the full .NET Framework, COM is already initialized as it is used by the framework itself. I am not sure how I can integrate this into SkiaSharp as this must be done outside the scope of `SKDocument`. the native skia has a nice utility type that I think I may be able to duplicate...

> https://msdn.microsoft.com/en-us/library/windows/desktop/ff485844.aspx

Here is a workaround:

```csharp
class Program
{
	static void Main(string[] args)
	{
		CoInitializeEx(IntPtr.Zero, COINIT.COINIT_APARTMENTTHREADED | COINIT.COINIT_DISABLE_OLE1DDE);

		using (var stream = new SKFileWStream("test.xps"))
		using (var document = SKDocument.CreateXps(stream))
		{
			Console.WriteLine($"document == '{document}'");

			document.BeginPage(100, 100);
			document.EndPage();
			document.Close();
		}

		Console.WriteLine("Hello World!");

		CoUninitialize();
	}

	public enum COINIT : uint
	{
		COINIT_MULTITHREADED = 0x0,
		COINIT_APARTMENTTHREADED = 0x2,
		COINIT_DISABLE_OLE1DDE = 0x4,
		COINIT_SPEED_OVER_MEMORY = 0x8,
	}

	[DllImport("ole32.dll", CharSet = CharSet.Auto, SetLastError = true, CallingConvention = CallingConvention.StdCall)]
	public static extern int CoInitializeEx([In, Optional] IntPtr pvReserved, [In]  COINIT dwCoInit);

	[DllImport("ole32.dll", CharSet = CharSet.Auto, SetLastError = true, CallingConvention = CallingConvention.StdCall)]
	public static extern void CoUninitialize();
}
```

**Comment by mattleibow** [comment]
The fix in v1.60.1 will be:

```csharp
using (new SKAutoCoInitialize()) {
   // use XPS bits
}
```

**NOTE: this is only needed for _some_ .NET Core apps. The full .NET Framework already initializes COM.**
